### PR TITLE
Query range fields by doc values when they are expected to be more efficient than points

### LINKED ISSUE #24314: Add doc values to range fields
If we add doc values to `range` fields and build doc-values-based range queries, then we could leverage `IndexOrDocValuesQuery` in order to use doc values when they are expected to be more efficient than points.

Related to #23182

### PR DESCRIPTION
**Pull request by martijnvg**
* Enable doc values for range fields by default.
* Store ranges in a binary format that support multi field fields.
* Added BinaryDocValuesRangeQuery that can query ranges that have been encoded into a binary doc values field.
* Wrap range queries on a range field in IndexOrDocValuesQuery query.

Closes #24314

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/mapper/RangeFieldMapper.java
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/index/mapper/RangeFieldMapper.java]: since we use a fixed length, maybe we do not need to encode the number of values?

Comment (jpountz) [inline core/src/main/java/org/elasticsearch/index/mapper/RangeFieldMapper.java]: oh I see it is to be consistent with other types which use vlong

**Review by jpountz — COMMENTED** [review]
It looks good overall. My main concern is that given how the encoding works, we have to deserialize the ranges at query time back to two `Comparable` instances. I think we should design it in such a way that queries can work directly on the encoded representation?

**Comment by martijnvg** [comment]
> I think we should design it in such a way that queries can work directly on the encoded representation?

The idea here is that we can save some space by encoding the ranges as two vlongs. Downside is that we can't work on the encoded representation. Do you think that the vlong encoding isn't worth the effort here? Personally I'm ok with changing the encoding here, so that we don't to encode at query time, but just wondering if we should make a tradeoff between space or query speed.

**Comment by jpountz** [comment]
You convinced me that we need to perform some sort of compression for numeric values. However I think we should use an encoding that doesn't require the query to know whether the field stores longs or ip addresses?

File: core/src/main/java/org/apache/lucene/queries/BinaryDocValuesRangeQuery.java
Comment (jpountz) [inline core/src/main/java/org/apache/lucene/queries/BinaryDocValuesRangeQuery.java]: This looks a bit backward to me, we would usually rather do
```
if (includeFrom == false) {
    from = rangeType.nextUp(from);
}
```
Do I miss something?

File: core/src/main/java/org/apache/lucene/queries/BinaryDocValuesRangeQuery.java
Comment (jpountz) [inline core/src/main/java/org/apache/lucene/queries/BinaryDocValuesRangeQuery.java]: the first bytes are not a vint when we encode eg. longs? So I don't think it works?

File: core/src/main/java/org/apache/lucene/queries/BinaryDocValuesRangeQuery.java
Comment (jpountz) [inline core/src/main/java/org/apache/lucene/queries/BinaryDocValuesRangeQuery.java]: We should try to avoid ByteArrayDataInput and BytesRef allocations since this method may be called in a tight loop.

File: core/src/main/java/org/apache/lucene/queries/BinaryDocValuesRangeQuery.java
Comment (jpountz) [inline core/src/main/java/org/apache/lucene/queries/BinaryDocValuesRangeQuery.java]: should we put a method on the queryType enum instead of this switch?

File: core/src/main/java/org/elasticsearch/index/mapper/RangeFieldMapper.java
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/index/mapper/RangeFieldMapper.java]: I'm wondering that putting these helpers into a dedicated class would make things easier to read?

File: core/src/main/java/org/elasticsearch/index/mapper/RangeFieldMapper.java
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/index/mapper/RangeFieldMapper.java]: I think we miss a mask here: `encoded[encoded.length - 1 - b] = (byte) (0xFF - ((l >>> (8 * b)) & 0xFF));`

File: core/src/test/java/org/apache/lucene/queries/LongRandomBinaryDocValuesRangeQueryTests.java
Comment (jpountz) [inline core/src/test/java/org/apache/lucene/queries/LongRandomBinaryDocValuesRangeQueryTests.java]: We had similar random number generation in eg. `TestIntRangeFieldQueries` but this proved inefficient at finding bugs (eg. it makes it very unlikely to get twice the same value in the same test) so we changed it in https://issues.apache.org/jira/browse/LUCENE-7847. Maybe we should do something similar here (as well as for doubles, ip addresses, etc.)

File: core/src/main/java/org/apache/lucene/queries/BinaryDocValuesRangeQuery.java
Comment (jpountz) [inline core/src/main/java/org/apache/lucene/queries/BinaryDocValuesRangeQuery.java]: Can we avoid passing the range type? I think it would be more robust if it does not need to know about the type of values that are stored and works directly on the encoded data?

**Review by jpountz — COMMENTED** [review]
I left some comments.

File: core/src/main/java/org/apache/lucene/queries/BinaryDocValuesRangeQuery.java
Comment (martijnvg) [inline core/src/main/java/org/apache/lucene/queries/BinaryDocValuesRangeQuery.java]: I can remove the ByteRef instance creation and use just plain byte[]. 
The reason I used `ByteArrayDataInput` is because of its `readVInt()` and `readBytes()` methods. Are there static alternatives that I can use? Otherwise maybe it is sufficient enough if I keep a spare instance around that I keep reusing?

**Comment by martijnvg** [comment]
@jpountz Thanks for reviewing. I've updated the PR.

File: core/src/main/java/org/elasticsearch/index/mapper/BinaryRangeUtil.java
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/index/mapper/BinaryRangeUtil.java]: doubles should probably use a different encoding, but we can work on it as a follow-up

Comment (jpountz) [inline core/src/main/java/org/elasticsearch/index/mapper/BinaryRangeUtil.java]: Also I think we should specialize floats too

File: core/src/main/java/org/elasticsearch/index/mapper/BinaryRangeUtil.java
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/index/mapper/BinaryRangeUtil.java]: Also I think we need some unit tests for that method in order to ensure it actually preserves ordering?

**Review by jpountz — COMMENTED** [review]
It looks good overall, I left some comments. I'm wondering whether we should sort the ranges before encoding them as it might allow for optimizations in the multi-valued case eg. if the max value of the query is less than the min value of the first encoded range.

It would be nice to have some benchmarks as well in order to make sure that this actually performs better than point-based ranges when the range does not drive iteration (it might make more sense to do it after [LUCENE-7828](https://issues.apache.org/jira/browse/LUCENE-7828) is merged). I can help with that.

**Comment by martijnvg** [comment]
@jpountz I've updated the pr.

>  I'm wondering whether we should sort the ranges before encoding them as it might allow for optimizations in the multi-valued case eg. if the max value of the query is less than the min value of the first encoded range.

I've added the sorting, but I think it will only make sense when there multiple adjacent ranges and when ranges are overlapping it wouldn't help that much?

> It would be nice to have some benchmarks as well in order to make sure that this actually performs better than point-based ranges when the range does not drive iteration (it might make more sense to do it after LUCENE-7828 is merged). I can help with that.

Thanks! I guess we will need a dataset that when benchmarked with a specific query forces the range query to be executed in random access mode.

**Comment by jpountz** [comment]
> I've added the sorting, but I think it will only make sense when there multiple adjacent ranges and when ranges are overlapping it wouldn't help that much?

Right I don't think this is a game changer for queries in general. However it might be the case for [aggregations](https://github.com/elastic/elasticsearch/issues/23182)? For instance, say you want to run an histogram aggregation on a range field, it would be easier with sorted ranges, since you can easily track which buckets have already been incremented and which buckets haven't?

**Comment by jpountz** [comment]
I merged LUCENE-7828 yesterday so that we can better benchmark this change, but upgrading to a new Lucene 7 snapshot is currently blocked on #25028 since the postings highlighter has been removed from Lucene master. It should hopefully be merged in the next few days.

**Comment by martijnvg** [comment]
This are the 99th percentile latency results between master (baseline) and this pr (contender) of the [new benchmark](https://github.com/elastic/rally-tracks/pull/30) added to rally:

|                        Metric |       Operation |   Baseline |   Contender |               Diff |   Unit |
| --------------------------- | ------------------  | ------------- | ---------------- |----------------- | ------- |
|       99th percentile latency |                                range_field_big_range |    4961.66 |     4643.11 | -318.556 |     ms |
|       99th percentile latency |                              range_field_small_range |    37.6339 |     29.3083 | -8.32565 |     ms |
|       99th percentile latency |   range_field_conjunction_big_range_small_term_query |     11.279 |     8.58807 |  -2.6909 |     ms |
|       99th percentile latency | range_field_conjunction_small_range_small_term_query |    6.15329 |     8.36214 |  2.20885 |     ms |
|       99th percentile latency |   range_field_conjunction_small_range_big_term_query |    2977.32 |     3478.42 |  501.095 |     ms |
|       99th percentile latency |     range_field_conjunction_big_range_big_term_query |    22320.1 |     21216.1 | -1104.07 |     ms |
|       99th percentile latency | range_field_disjunction_small_range_small_term_query |    1438.64 |     45.5729 | -1393.07 |     ms |
|       99th percentile latency |   range_field_disjunction_big_range_small_term_query |    8843.91 |      5680.8 | -3163.11 |     ms |

**Review by jpountz — CHANGES_REQUESTED** [review]
I think we need to do 2 things before merging:
 - back compat in the mapper so that the pre-6.0 mappers know they do not have doc values
 - sort ranges by from then to

**Comment by martijnvg** [comment]
@jpountz I've updated the pr.

**Review by jpountz — APPROVED** [review]

**Pull request merged**
